### PR TITLE
python: implement stream-related classes

### PR DESCRIPTION
**Pull request by crockeo**
**Description:**

All header changes in this PR come from https://github.com/envoyproxy/envoy-mobile/pull/1217, that should be reviewed and merged first.

This PR implements all of the stream-related classes: `Stream`, `StreamCallbacks `EnvoyHttpCallbacksAdapter`, and `StreamPrototype`.

**Risk Level:** Low
**Testing:** N/A, will follow up when everything is finished with e2e testing + unit testing in Python and/orC++
**Docs Changes:** N/A
**Release Notes:** N/A

### CODE REVIEW

**Comment by goaway** [comment]
Can you update this (and others) to track #1217 to make it easier to see what the diff is?

**Comment by crockeo** [comment]
I don't actually think I can, since I've forked the repo 🙁 

What I could do is reopen these PRs against my fork and then set them to track b6a6519 there?

File: library/cc/stream.cc [hunk 1]
```diff
@@ -0,0 +1,37 @@
...
+#include "library/common/main_interface.h"
+#include "library/common/types/c_types.h"
+
+namespace Envoy {
+namespace Platform {
+
+Stream::Stream(envoy_stream_t handle, EnvoyHttpCallbacksAdapterSharedPtr adapter)
+    : handle_(handle), adapter_(adapter) {}
+
+Stream& Stream::send_headers(RequestHeadersSharedPtr headers, bool end_stream) {
```
Comment (goaway) [inline library/cc/stream.cc:15]: Envoy style is for all instance methods to be camelCase.

File: library/cc/stream_callbacks.cc [hunk 2]
```diff
@@ -0,0 +1,96 @@
...
+      .on_trailers = &EnvoyHttpCallbacksAdapter::c_on_trailers,
+      .on_error = &EnvoyHttpCallbacksAdapter::c_on_error,
+      .on_complete = &EnvoyHttpCallbacksAdapter::c_on_complete,
+      .on_cancel = &EnvoyHttpCallbacksAdapter::c_on_cancel,
+      .context = this,
+  };
+  return callbacks;
+}
+
+void* EnvoyHttpCallbacksAdapter::c_on_headers(envoy_headers headers, bool end_stream,
```
Comment (goaway) [inline library/cc/stream_callbacks.cc:27]: These though I think we can leave as snake_case to distinguish them

File: library/cc/stream.cc (hunk 1 above)
Comment (goaway) [reply library/cc/stream.cc:15]: Let's worry about this later though, I'm going to approve and merge.

**Review by goaway — APPROVED** [review]

**Pull request merged**